File: HLD/Databases/code/indexing.py
import random
import time

NUM_RECORDS = 200_000
NUM_LOOKUPS = 500

def build_table(n):
    return [{"id": i, "name": f"user-{i}"} for i in range(n)]

def build_index(table):
    return {row["id"]: row for row in table}

def linear_scan_lookup(table, target_id):
    for row in table:
        if row["id"] == target_id:
            return row
    return None

def indexed_lookup(index, target_id):
    return index.get(target_id)

if __name__ == "__main__":
    table = build_table(NUM_RECORDS)
    index = build_index(table)
    lookup_ids = [random.randint(0, NUM_RECORDS - 1) for _ in range(NUM_LOOKUPS)]

    start = time.perf_counter()
    for target in lookup_ids:
        linear_scan_lookup(table, target)
    linear_time_ms = (time.perf_counter() - start) * 1000

    start = time.perf_counter()
    for target in lookup_ids:
        indexed_lookup(index, target)
    indexed_time_ms = (time.perf_counter() - start) * 1000

    print(f"{NUM_RECORDS} records, {NUM_LOOKUPS} lookups")
    print(f"Linear scan (no index): {linear_time_ms:.2f} ms")
    print(f"Indexed lookup (hash index): {indexed_time_ms:.2f} ms")
    print(f"Speedup from indexing: {linear_time_ms / indexed_time_ms:.1f}x")

File: HLD/Databases/code/indexing.js
const NUM_RECORDS = 200000;
const NUM_LOOKUPS = 500;

function buildTable(n) {
  const table = new Array(n);
  for (let i = 0; i < n; i++) {
    table[i] = { id: i, name: `user-${i}` };
  }
  return table;
}

function buildIndex(table) {
  const index = new Map();
  for (const row of table) {
    index.set(row.id, row);
  }
  return index;
}

function linearScanLookup(table, targetId) {
  for (const row of table) {
    if (row.id === targetId) return row;
  }
  return null;
}

function indexedLookup(index, targetId) {
  return index.get(targetId) || null;
}

const table = buildTable(NUM_RECORDS);
const index = buildIndex(table);
const lookupIds = Array.from({ length: NUM_LOOKUPS }, () => Math.floor(Math.random() * NUM_RECORDS));

let start = process.hrtime.bigint();
for (const target of lookupIds) {
  linearScanLookup(table, target);
}
const linearTimeMs = Number(process.hrtime.bigint() - start) / 1e6;

start = process.hrtime.bigint();
for (const target of lookupIds) {
  indexedLookup(index, target);
}
const indexedTimeMs = Number(process.hrtime.bigint() - start) / 1e6;

console.log(`${NUM_RECORDS} records, ${NUM_LOOKUPS} lookups`);
console.log(`Linear scan (no index): ${linearTimeMs.toFixed(2)} ms`);
console.log(`Indexed lookup (hash index): ${indexedTimeMs.toFixed(2)} ms`);
console.log(`Speedup from indexing: ${(linearTimeMs / indexedTimeMs).toFixed(1)}x`);
